distribution/std-lib/Database/src/Connection/Connection.enso

```diff
@@ -23,7 +35,182 @@ type Connection
 
        Arguments:
        - name: name of the table to access
-    access_table : Text -> Table
-    access_table name =
-        column_names = this.internal_connection.fetch_column_names name
-        Table.make_table this name column_names
+    access_table : Text -> Database_Table
+    access_table name = here.wrap_sql_errors <|
+        columns = this.fetch_columns name
+        Database_Table.make_table this name columns
+
+    ## Closes the connection releasing the underlying database resources
+       immediately instead of waiting for them to be automatically released.
+
+       The connection is not usable afterwards.
+    close : Nothing
+    close =
+        Managed_Resource.finalize this.connection_resource
+
+    ## ADVANCED
+
+       Executes a raw query and returns the result as an in-memory Table.
+
+       Arguments:
+       - query: either raw SQL code as Text or an instance of Sql.Statement
+         representing the query to execute.
+       - expected_types: an optional array of expected types of each column;
+         meant only for internal use.
+    execute_query : Text | Sql.Statement -> Vector Sql.Sql_Type -> Materialized_Table =
+    execute_query query expected_types=Nothing = here.wrap_sql_errors <|
+        Resource.bracket (this.prepare_statement query) .close stmt->
+            rs = stmt.executeQuery
+            metadata = rs.getMetaData
+            ncols = metadata.getColumnCount
+            ## TODO [RW] we wrap everything in objects, in the future we may
+               want to try a more performant solution for columns of numbers or
+               booleans
+            column_names = Vector.new ncols (ix -> metadata.getColumnName ix+1)
+            column_types = if expected_types.is_nothing.not then expected_types else
+                Vector.new ncols (ix -> Sql.Sql_Type <| metadata.getColumnType ix+1)
+            columns = column_names.map (name -> [name, ArrayList.new])
+            go has_next = if has_next.not then Nothing else
+                columns.map_with_index ix-> pair->
+                    results = pair.second
+                    typ = column_types.at ix
+                    ## getObject handles most types by default correctly, but it
+                       would cast booleans to integers
+                    obj = case typ == Sql.Sql_Type.boolean of
+                        True ->
+                            bool = rs.getBoolean ix+1
+                            if rs.wasNull then Nothing else bool
+                        False -> rs.getObject ix+1
+                    results.add obj
+                go rs.next
+            go rs.next
+            column_vectors = columns.map (pair -> [pair.first, Vector.Vector pair.second.toArray])
+            Materialized_Table.new column_vectors
+
+    ## ADVANCED
+
+       Executes a raw update query and returns the amount of affected rows (or 0
+       if not applicable).
```

Comment:
What I wanted to convey here is that 0 may mean that the operation is of other type (for example `CREATE TABLE`, `ALTER` etc. will return 0) and to be precise `ALTER TABLE` may alter rows but afaik it will return 0 too.

Maybe we could phrase this as:
```
Executes a raw update query. If the query was inserting, updating or deleting rows, the number of affected rows is returned; otherwise it returns 0 for other types of queries (like creating or altering tables).
```

What do you think?